plugins/parade-lspcon/fu-parade-lspcon-device.c

```diff
@@ -0,0 +1,829 @@
+	out = g_new0 (FuParadeLspconI2cAddressGuard, 1);
+	out->device = self;
+	return out;
+}
+
+static void
+fu_parade_lspcon_i2c_address_guard_free (FuParadeLspconI2cAddressGuard *guard)
+{
+	g_autoptr(GError) error_local = NULL;
+	if (!fu_parade_lspcon_ensure_i2c_address (guard->device,
+						  I2C_ADDR_PAGE2,
+						  &error_local)) {
+		g_warning ("failed to set page2 back: %s", error_local->message);
+	}
+	g_free (guard);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(FuParadeLspconI2cAddressGuard,
+			      fu_parade_lspcon_i2c_address_guard_free);
+
+static gboolean
+fu_parade_lspcon_write_register (FuParadeLspconDevice *self,
+				 guint8 register_addr,
+				 guint8 value,
+				 GError **error)
+{
+	guint8 transaction[] = { register_addr, value };
+	return fu_i2c_device_write_full (FU_I2C_DEVICE (self), transaction,
+					 sizeof(transaction), error);
+}
+
+static gboolean
+fu_parade_lspcon_read_register (FuParadeLspconDevice *self,
+				guint8 register_addr,
+				guint8 *value,
+				GError **error)
+{
+	FuI2cDevice *i2c_device = FU_I2C_DEVICE (self);
+	if (!fu_i2c_device_write (i2c_device, register_addr, error))
+		return FALSE;
+	return fu_i2c_device_read (i2c_device, value, error);
+}
+
+/* map the page containing the given address into page 7 */
+static gboolean
+fu_parade_lspcon_map_page (FuParadeLspconDevice *self, guint32 address, GError **error)
+{
+	if (!fu_parade_lspcon_write_register (self,
+					      REG_ADDR_FLASH_ADDR_HI,
+					      address >> 16,
+					      error))
+		return FALSE;
+	return fu_parade_lspcon_write_register (self,
+						REG_ADDR_FLASH_ADDR_LO,
+						address >> 8,
+						error);
+}
+
+/* wait until the specified register masked with mask reads the expected
+ * value, up to 10 seconds */
+static gboolean
+fu_parade_lspcon_poll_register (FuParadeLspconDevice *self,
+				guint8 register_address,
+				guint8 mask,
+				guint8 expected,
+				GError **error)
+{
+	guint8 value;
+	g_autoptr(GTimer) timer = g_timer_new();
+
+	do {
+		if (!fu_parade_lspcon_read_register (self,
+						     register_address,
+						     &value,
+						     error))
+			return FALSE;
+		if ((value & mask) == expected)
+			return TRUE;
+	} while (g_timer_elapsed (timer, NULL) <= 10.0);
+
+	g_set_error (error,
+		     G_IO_ERROR,
+		     G_IO_ERROR_TIMED_OUT,
+		     "register %x did not read %x (mask %x) within 10 seconds: read %x",
+		     register_address, expected, mask, value);
+	return FALSE;
+}
+
+static gboolean
+fu_parade_lspcon_flash_read (FuParadeLspconDevice *self,
+			     guint32 base_address,
+			     guint8 *data,
+			     gsize len,
+			     GError **error)
+{
+	FuDevice *device = FU_DEVICE (self);
+	FuI2cDevice *i2c_device = FU_I2C_DEVICE (self);
+	const gsize target_len = len;
+
+	while (len > 0) {
+		/* page 7 reads always start from the base of the mapped window- we'll
+		 * read the whole page then pull out the parts we care about, using the
+		 * full page everywhere except possibly in the first and last reads */
```

Comment:
Does FuChunk's `fu_chunk_array_new` help here? You can set the page size in that.

Reply:
We don't have a way to read anywhere but from the beginning of a page, so chunking won't help us significantly here.